### PR TITLE
Simplify VNet setup

### PR DESCRIPTION
**Pull request by ravicious**
After [a conversation with Alan](https://gravitational.slack.com/archives/C0DF0TPMY/p1715853418962919), I decided to try out a couple of improvements which could simplify what I've done in #41413.

In #41413, `vnet.Setup` and `vnet.Run` serve mostly as a big bags of stray functions that are needed to start VNet that I wanted to share between tsh and Connect. The main problem the split into `Setup` and `Run` was solving is that tsh can set up VNet and then start it while blocking the main thread, whereas Connect cannot do this. The RPC in Connect that sets up VNet needs to return after a successful setup and let the VNet manager and the admin subcommand run in the background.

This PR:

* Explicitly throws blocking calls into `errgroup` to get rid of all the shenanigans I was doing with contexts in #41413.
* Removes aggregation of errors from running the admin subcommand and VNet manager. If one fails we cancel the other and forward only the first error. This gets rid of a bunch of different shenanigans with `cancel`.

The setup code ends up being a little bit more elaborate, but I think it's easier to follow and doesn't raise as much red flags as the one I introduced in #41413 does.

### CODE REVIEW

File: tool/tsh/common/vnet_darwin.go
Comment (nklaassen) [inline tool/tsh/common/vnet_darwin.go:86]: to be honest the exported API is still a bit ugly to deal with in both tsh and connect

I think ideally for clients it would look something like this:

```go
// tsh
vnetManager, err := vnet.Setup(ctx, appProvider)
if err != nil {
  return trace.Wrap(err)
}
return trace.Wrap(vnetManager.Run(ctx))
```

```go
// connect
vnetManager, err := vnet.Setup(ctx, appProvider)
if err != nil {
  return trace.Wrap(err)
}
vnetRunCtx, vnetCancel := context.WithCancel(context.Background())
go func() {
  err := vnetManager.Run(vnetRunCtx)
  // TODO: report error or early termination somehow
}()
```

The admin subcommand would be started in `vnet.Setup` and get to the point where the password prompt completed succesfully and we got the TUN, that's everything you might want to cancel in the request context. From that point managing the lifecycle of the admin process would be handled within the `vnetManager`, which would also hide implementation details like the socket. I think a simplified exported API like that would be much nicer if it's doable.

the downside is it's stuffing more things into `vnet.Manager` (which i know was not a good name to start with). We could either do that, or rename the current `vnet.Manager` to something like `vnet.NetworkStack`, and make a new `vnet.ProcessManager` that manages the lifetime of the admin process as well as the `NetworkStack`. There would be some options there but I think that's less important than getting the exported API right

let me know if I'm making sense

Comment (ravicious) [reply tool/tsh/common/vnet_darwin.go:86]: I like the idea of the introducing `NetworkStack` and `ProcessManager`. The proposed API does look nice, but I don't think it's possible to have it look exactly like this. `vnet.Setup` in Connect cannot take `ctx`, because that context is bound to the RPC lifetime and it would stop the admin process right after the RPC ends. The admin process needs to be executed with a long-running context.

But this gives me an idea, what if `vnetManager.Run` just didn't accept a context and we followed the API of something like `net/http.Server`?

We could have something like this:

```go
// connect
// Still accepts ctx – it's needed when waiting for the TUN device to be created.
// But the admin subcmd and other stuff is spawned using a background context
// that's not exposed outside of processManager.
success := false
processManager, err := vnet.Setup(ctx, appProvider)
if err != nil {
  return trace.Wrap(err)
}
defer func() {
  if !success {
    err := processManager.Close()
    // TODO: Log the error.
  }
}()
s.processManager = processManager
success = true

go func() {
  err := processManager.Run()
  // TODO: report error or early termination somehow
}()



// In the Stop RPC
s.processManager.Close()
```

I'll try to explore this idea, though idk yet if I'm going to branch off this PR or #41413. I like the API from this PR slightly more than the previous one, mostly because it clearly separates the long running stuff from other operations, without having to pass `adminCommandErrCh` around, etc.

Comment (ravicious) [reply tool/tsh/common/vnet_darwin.go:86]: The only difference when comparing to something like `net/http.Server` would be that the caller is expected to call `processManager.Close()` not only after `processManager.Run`, but also after `vnet.Setup`. But I believe we can address this either through comments or function names.

Comment (ravicious) [reply tool/tsh/common/vnet_darwin.go]: Alright, I cleaned it up, let me know what you think. `vnet.ProcessManager` ended up being an `errgroup.Group` on wheels, maybe we could even name it `TaskGroup` or something.

The biggest hiccup is around handling the cancellation of osascript prompt. The error we want to show to the user comes from `execAdminSubcommand`. We somehow have to make sure that the user sees that before the error about reading from a closed socket.

https://github.com/gravitational/teleport/blob/0cff6ea1f979ab5298c34b6449b7b19dc191ac13/lib/vnet/setup.go#L70-L106

File: lib/teleterm/vnet/service.go
Comment (nklaassen) [reply lib/teleterm/vnet/service.go:177]: oh yeah I see now that `processManager.Close()` also waits

File: lib/vnet/vnet.go
Comment (nklaassen) [inline lib/vnet/vnet.go:158]: there's a bunch of things we probably don't need to export anymore, but don't worry about that in this PR, we should probably just clean that up at some point

**Comment by nklaassen** [comment]
i had a bunch of implementation suggestions but overall i do like this much better now

File: lib/teleterm/vnet/service.go
Comment (ravicious) [reply lib/teleterm/vnet/service.go:111]: Right, I forgot to remove this after I moved pretty much all setup stuff into `vnet.SetupAndRun`.

File: lib/vnet/setup.go
Comment (ravicious) [reply lib/vnet/setup.go:152]: > I think this has better semantics if we promise to cancel all tasks if a single task exits at all, rather then just when it exits with an error

Definitely. I suppose I could remove this fragment from `execAdminSubcommand`, but OTOH I like that it gives more information about the problem.

https://github.com/gravitational/teleport/blob/96f32a23e893c8168df0b9ba273dcc1b18d2fb15/lib/vnet/setup_darwin.go#L105-L117

File: lib/vnet/setup.go
Comment (ravicious) [reply lib/vnet/setup.go:50]: I do know that it's a best practice to not store a context, idk why I thought I could get away with it here. It should also simplify closing the socket – it doesn't need to be done from a separate `AddCriticalBackgroundTask` call.

If we don't store the context, then in `AddCriticalBackgroundTask` we cannot do this:

```go
		err := task(pm.ctx)
		if err == nil && pm.ctx.Err() == nil {
			err = fmt.Errorf("critical task %q exited prematurely", name)
		}
```

But I don't think that check on `ctx` is even necessary. We should just expect all tasks to always return with an error (be it `context.Canceled` or any other error). Swallowing `context.Canceled` should be handled at the outermost layer.

File: lib/vnet/setup.go
Comment (ravicious) [reply lib/vnet/setup.go:55]: If we make it so that `Close` does not `Wait` underneath, would it be fine to just call `Close` here and not `Wait` for any errors?

Comment (ravicious) [reply lib/vnet/setup.go:55]: In 99% of cases the errors from `Wait` would be `context.Canceled` anyway, which I suppose we don't want to propagate if `SetupAndRun` because, for example, a socket couldn't have been created.

File: lib/vnet/vnet.go
Comment (ravicious) [reply lib/vnet/vnet.go:664]: Yeah I also screwed up the capitalization here. 😶

File: lib/vnet/setup.go
Comment (ravicious) [reply lib/vnet/setup.go:55]: I guess technically those two background tasks might react differently to a context being canceled and one could just forward `context.Canceled`, while the other would run into another problem while closing. Which I suppose was the original intention behind capturing all errors from `errgroup` and not just the first one, to avoid a race. 😣

Comment (ravicious) [reply lib/vnet/setup.go:55]: But in this specific scenario it shouldn't matter (I think) because `SetupAndRun` has failed anyway and we just want to clean up any resources.

File: lib/vnet/setup.go
Comment (ravicious) [reply lib/vnet/setup.go]: This works, but only if we replace `processCtx.Err()` with `context.Cause(processCtx)`. Otherwise if the password prompt gets closed, we're going to bubble up `context.Canceled` instead of the actual error about the password prompt getting canceled.

However, now I ran into another problem. When interrupting `tsh vnet` after the TUN device is set up, sometimes I get an error about the TUN device being already closed. I'm yet to pinpoint what causes this.

```
$ tsh vnet -d
2024-05-20T13:20:28+02:00 DEBU  Created unix socket for admin subcommand socket:/var/folders/kz/2d_vhlks21l3g0s4g5yytlwr0000gn/T/vnet7b3e4f1b-f249-479e-823f-c087d5a72942.sock trace_id:dab3f5c9c15375df6acdb8a7f8a5af17 span_id:3f9a786e64b9d9e6 vnet/setup.go:65
2024-05-20T13:20:31+02:00 DEBU [VNET]      Serving DNS on IPv6. dns_addr:fdf4:cad0:45f4::2 vnet/vnet.go:266
2024-05-20T13:20:31+02:00 INFO [VNET]      Running Teleport VNet. ipv6_prefix:fdf4:cad0:45f4:: vnet/vnet.go:306
2024-05-20T13:20:31+02:00 DEBU  Forwarding IP packets between OS and VNet. vnet/vnet.go:572
^C
ERROR REPORT:
Original Error: trace.aggregate reading packets from TUN
	read utun4: file already closed
Stack Trace:
	github.com/gravitational/trace@v1.4.0/trace.go:471 github.com/gravitational/trace.NewAggregateFromChannel
	github.com/gravitational/teleport/lib/vnet/vnet.go:346 github.com/gravitational/teleport/lib/vnet.(*NetworkStack).Run
	github.com/gravitational/teleport/lib/vnet/setup.go:119 github.com/gravitational/teleport/lib/vnet.SetupAndRun.func5
	github.com/gravitational/teleport/lib/vnet/setup.go:149 github.com/gravitational/teleport/lib/vnet.SetupAndRun.(*ProcessManager).AddCriticalBackgroundTask.func7
	golang.org/x/sync@v0.7.0/errgroup/errgroup.go:78 golang.org/x/sync/errgroup.(*Group).Go.func1
	runtime/asm_arm64.s:1222 runtime.goexit
User Message: reading packets from TUN
	read utun4: file already closed
```

Comment (ravicious) [reply lib/vnet/setup.go]: FWIW, I can reproduce this on a version from before I started refactoring stuff today (96f32a23e8) and before I introduced ProcessManager (4b54ab948e6114907e).

I cannot reproduce it on 3984575a6c, which is https://github.com/gravitational/teleport/pull/41413#event-12823441001 from before rebase on `nklaassen/vnet5` from a few days ago.

Comment (ravicious) [reply lib/vnet/setup.go]: Oh I see, `forwardTUNtoNetstack` does not receive a context, so it just gets stuck on `tun.Read`. If it returns before `forwardNetstackToTUN` manages to return `context.Canceled`, then we bubble up `file already closed` instead of `context.Canceled`.

Comment (ravicious) [reply lib/vnet/setup.go]: I'm not sure how to address that though.

Comment (ravicious) [reply lib/vnet/setup.go]: I was able to consistently reproduce it by artificially slowing down `forwardNetstackToTUN` and `execAdminSubcommand` so that `ns.Run` manages to always be the first to return an error due to `forwardTUNtoNetstack` failing with `os.ErrClosed`.

I googled around and found a solution to this with checking `ctx.Err` before returning an error from `file.Read`. I'm not sure how idiomatic it is though: 247450b7

File: lib/vnet/setup.go
Comment (zmb3) [inline lib/vnet/setup.go:139]: Took me a few minutes and some poking around to fully grok this, but it's not too bad once you understand it.

I'd recommend writing a few tests for it that execute arbitrary tasks which are otherwise unrelated to VNet. Had those already existed I may have learned enough just by looking at the tests rather than having to write my own code to exercise this.

**Comment by ravicious** [comment]
Didn't add anything new today besides the tests, just rebasing on top of my custom branch off of `nklaassen/vnet6` to have an easier time merging.

File: lib/vnet/setup.go
Comment (nklaassen) [reply lib/vnet/setup.go:53]: with Close not returning any error now i think this is fine

File: lib/vnet/setup.go
Comment (nklaassen) [reply lib/vnet/setup.go]: oh yeah sorry for the conflicting comments, guess i didn't fully think that through, it think what you have now works

**Comment by ravicious** [comment]
@nklaassen After I rebased it on top of `nklaassen/vnet6`, something went wrong with terminating the admin subcommand and now it leaves leftovers in `/etc/resolver`. 😭

I think it's maybe because `nklaassen/vnet6` is outdated or something, I swear it used to work when I had it rebased on top of vnet5 in 673bb43.

I'm signing off today, so I'll try to hunt it down tomorrow.

File: lib/vnet/setup.go
Comment (nklaassen) [reply lib/vnet/setup.go]: yeah i think it's probably still good to leave that

**Comment by ravicious** [comment]
But generally it'd be nice if I could get this PR through reviews by the end of tomorrow, then I'd have Thursday and Friday to get other dependent PRs through.

**Comment by nklaassen** [comment]
> @nklaassen After I rebased it on top of `nklaassen/vnet6`, something went wrong with terminating the admin subcommand and now it leaves leftovers in `/etc/resolver`. 😭
> 
> I think it's maybe because `nklaassen/vnet6` is outdated or something, I swear it used to work when I had it rebased on top of vnet5 in [673bb43](https://github.com/gravitational/teleport/commit/673bb4333ffbedf9c18eb9908e0302b1a2815112).
> 
> I'm signing off today, so I'll try to hunt it down tomorrow.

This is the weirdest thing... I had the osascript shell command redirecting logs to a file for debugging, and when I removed that, the /etc/resolver cleanup broke. But I can't debug it without logs 😭 so I added back the log redirect to `/var/log/vnet.log` and now it's working again

File: lib/vnet/setup.go
Comment (ravicious) [reply lib/vnet/setup.go:139]: Tests added in 3bbcc2a2c37829.

**Comment by ravicious** [comment]
@nklaassen @zmb3 Are there any other outstanding issues that I should address? It'd be great if I could split this tomorrow and merge it into #41413 and #41415.

File: lib/vnet/setup.go
Comment (nklaassen) [reply lib/vnet/setup.go]: that works

**Review by nklaassen — APPROVED** [review]

**Comment by nklaassen** [comment]
> @nklaassen @zmb3 Are there any other outstanding issues that I should address? It'd be great if I could split this tomorrow and merge it into #41413 and #41415.

no it seems good to me, a bit hard to review on top of https://github.com/gravitational/teleport/pull/41413, i'll make sure to review after you merge it in there

**Review by zmb3 — APPROVED** [review]

**Comment by ravicious** [comment]
Thanks for the reviews. I squashed `lib/vnet` and `tool/tsh` code from here into #41413 and `lib/teleterm` into #41415.

**Pull request closed without merging**
